# Unit 3: 輸入輸出與格式化 | Input/Output and Formatting

## 📖 講義 | Lecture Notes

---

## Part I: 理論基礎 | Theoretical Foundations

### 📚 章節概覽（Chapter Overview）

#### 學習目標（Learning Objectives）
完成本章後，您將能夠：
1. 掌握 print() 與 input() 函式的使用
2. 理解 input() 回傳字串的特性
3. 使用 f-string 進行字串格式化
4. 建立與使用者互動的程式

#### 先備知識（Prerequisites）
- Unit 1: 變數與資料型態
- Unit 2: 運算子與表達式



---

### 🔑 核心概念（Key Concepts）

#### 1. 什麼是輸入輸出？（What is I/O?）

**定義**：
- **輸出（Output）**：程式將資料顯示給使用者
- **輸入（Input）**：程式從使用者獲取資料

**類比**：
```
程式與使用者的對話：
程式: "請輸入你的名字" → print()
使用者: "Alice"        → input()
程式: "你好，Alice！"  → print()
```

#### 2. 為什麼需要格式化？（Why Formatting?）

**根本問題**：原始資料輸出不易閱讀

**比較**：
```python
# 未格式化（難讀）
print(3.14159265359)  # 太多小數

# 格式化（易讀）
print(f"{3.14159265359:.2f}")  # 3.14
```

---

## Part II: 實作演練 | Hands-on Practice

### 💡 範例 1：print() 基本用法

最基本的輸出功能

In [2]:
# 基本輸出
print("Hello, World!")

# 輸出變數
name = "Alice"
age = 25
print(name)
print(age)

# 輸出多個值（用空格分隔）
print("姓名:", name, "年齡:", age)

Hello, World!
Alice
25
姓名: Alice 年齡: 25


**關鍵要點**：
- print() 可接受任意數量的參數
- 預設用空格分隔，結尾換行

---

### 💡 範例 2：print() 的 sep 與 end 參數

In [3]:
# sep 參數：自訂分隔符
print("2025", "10", "05", sep="-")  # 2025-10-05
print("Apple", "Banana", "Cherry", sep=", ")  # Apple, Banana, Cherry

# end 參數：自訂結尾字元
print("第一行", end=" ")  # 不換行，改用空格
print("第二行")  # 結果：第一行 第二行

# 組合使用
print("A", "B", "C", sep="-", end="!\n")  # A-B-C!

2025-10-05
Apple, Banana, Cherry
第一行 第二行
A-B-C!


**實際應用**：

In [4]:
# 應用：進度條效果
import time
for i in range(5):
    print("#", end="", flush=True)
    time.sleep(0.5)
print(" 完成！")

##### 完成！


---

### 💡 範例 3：input() 基本用法

從使用者獲取輸入

In [5]:
# 基本輸入（顯示提示訊息）
name = input("請輸入你的名字：")
print(f"你好，{name}！")

# 檢查 input() 的回傳型態
user_input = input("請輸入一個數字：")
print(f"你輸入的是：{user_input}")
print(f"型態是：{type(user_input)}")  # <class 'str'>

你好，Sunny！
你輸入的是：50
型態是：<class 'str'>


**重要提醒**：
- input() **永遠**回傳 str
- 即使使用者輸入數字，也是字串格式

---

### 💡 範例 4：輸入的型態轉換

將字串輸入轉換為數字

In [6]:
# 獲取整數
age_str = input("請輸入年齡：")
age = int(age_str)  # 字串轉整數
print(f"你今年 {age} 歲，明年 {age + 1} 歲")

# 簡化寫法（直接轉換）
height = float(input("請輸入身高(cm)："))
print(f"你的身高是 {height} cm")

你今年 18 歲，明年 19 歲
你的身高是 165.0 cm


**常見錯誤**：

In [7]:
# ❌ 錯誤：忘記轉換型態
# num1 = input("數字1：")  # "5" (str)
# num2 = input("數字2：")  # "3" (str)
# result = num1 + num2     # "53" (字串串接，不是 8)

# ✅ 正確：先轉換再運算
num1 = int(input("數字1："))
num2 = int(input("數字2："))
result = num1 + num2  # 8 (數字相加)
print(f"相加結果：{result}")

相加結果：10


---

### 💡 範例 5：f-string 格式化（推薦）

Python 3.6+ 的現代格式化方法

In [8]:
# 基本 f-string
name = "Bob"
age = 30
height = 175.5

# 變數插入
print(f"我是 {name}，今年 {age} 歲")

# 運算式插入
print(f"明年我 {age + 1} 歲")
print(f"身高: {height / 100} 公尺")

# 方法呼叫
print(f"大寫名字: {name.upper()}")

我是 Bob，今年 30 歲
明年我 31 歲
身高: 1.755 公尺
大寫名字: BOB


---

### 💡 範例 6：數字格式化

控制數字的顯示格式

In [9]:
pi = 3.141592653589793

# 控制小數位數
print(f"預設: {pi}")
print(f"2位小數: {pi:.2f}")
print(f"4位小數: {pi:.4f}")

# 千分位
large_num = 1234567.89
print(f"千分位: {large_num:,.2f}")  # 1,234,567.89

# 百分比
ratio = 0.856
print(f"百分比: {ratio:.1%}")  # 85.6%

預設: 3.141592653589793
2位小數: 3.14
4位小數: 3.1416
千分位: 1,234,567.89
百分比: 85.6%


---

### 💡 範例 7：對齊與填充

製作整齊的表格輸出

In [10]:
# 右對齊（數字常用）
print(f"{42:>10}")    # '        42'
print(f"{123:>10}")   # '       123'

# 左對齊（文字常用）
print(f"{'Alice':<10}end")   # 'Alice     end'
print(f"{'Bob':<10}end")     # 'Bob       end'

# 置中對齊
print(f"{'Title':^20}")  # '       Title       '

# 補零（數字）
print(f"{42:05d}")       # '00042'
print(f"{7:03d}")        # '007'

        42
       123
Alice     end
Bob       end
       Title        
00042
007


**實際應用：格式化表格**

In [11]:
# 輸出對齊的表格
print("學生成績表")
print("-" * 30)
print(f"{'姓名':<10} {'成績':>8} {'等級':>8}")
print("-" * 30)
print(f"{'Alice':<10} {95:>8d} {'A':>8}")
print(f"{'Bob':<10} {82:>8d} {'B':>8}")
print(f"{'Charlie':<10} {78:>8d} {'C':>8}")
print("-" * 30)

學生成績表
------------------------------
姓名               成績       等級
------------------------------
Alice            95        A
Bob              82        B
Charlie          78        C
------------------------------


---

### 💡 範例 8：完整的互動程式

結合 input、運算、格式化輸出

In [12]:
# BMI 計算程式（互動版）
print("=== BMI 計算器 ===")

# 步驟 1：獲取輸入
name = input("請輸入姓名：")
weight = float(input("請輸入體重(kg)："))
height_cm = float(input("請輸入身高(cm)："))

# 步驟 2：計算 BMI
height_m = height_cm / 100
bmi = weight / (height_m ** 2)

# 步驟 3：判斷健康狀況
if bmi < 18.5:
    status = "體重過輕"
elif bmi < 24:
    status = "正常範圍"
elif bmi < 27:
    status = "過重"
else:
    status = "肥胖"

# 步驟 4：格式化輸出結果
print(f"\n=== {name} 的 BMI 報告 ===")
print(f"體重：{weight:.1f} kg")
print(f"身高：{height_cm:.1f} cm")
print(f"BMI：{bmi:.2f}")
print(f"狀況：{status}")

=== BMI 計算器 ===

=== Sunny 的 BMI 報告 ===
體重：62.0 kg
身高：173.0 cm
BMI：20.72
狀況：正常範圍


---

## Part III: 本章總結 | Chapter Summary

### 📊 知識回顧

#### 核心概念
1. **print()**：輸出資料給使用者
   - 參數：`sep`（分隔符）、`end`（結尾）
2. **input()**：從使用者獲取資料
   - 永遠回傳 str
   - 需手動轉換型態
3. **f-string**：格式化字串（推薦）
   - 語法：`f"{變數:格式規範}"`

#### 重要語法
```python
# 輸出
print(value1, value2, sep=" ", end="\n")

# 輸入
user_input = input("提示訊息：")

# 格式化
f"{變數}"
f"{變數:.2f}"      # 2位小數
f"{變數:>10}"      # 右對齊寬度10
f"{變數:05d}"      # 補零寬度5
```

#### 常見格式規範

| 規範 | 說明 | 範例輸出 |
|:-----|:-----|:---------|
| `:.2f` | 浮點數2位小數 | 3.14 |
| `:,` | 千分位 | 1,000 |
| `:>10` | 右對齊寬度10 | "       123" |
| `:<10` | 左對齊寬度10 | "123       " |
| `:^10` | 置中寬度10 | "   123    " |
| `:05d` | 補零寬度5 | 00123 |
| `:.1%` | 百分比1位小數 | 85.6% |

### ⚠️ 常見誤區（Common Pitfalls）

| 誤區 | 錯誤示例 | 正確做法 |
|:-----|:---------|:---------|
| 忘記轉換 input() | `age = input()` 直接運算 | `age = int(input())` |
| 字串與數字串接 | `"年齡:" + 25` | `f"年齡:{25}"` 或 `"年齡:" + str(25)` |
| 格式化語法錯誤 | `f"{pi.2f}"` | `f"{pi:.2f}"` (注意冒號) |
| 忘記 f 前綴 | `"{name}"` | `f"{name}"` |

---

### 🎯 自我檢核（Self-Check）

完成本講義後，請回答以下問題：

1. print() 的預設分隔符是什麼？
2. input() 回傳的型態永遠是什麼？
3. `f"{3.14159:.2f}"` 的輸出是什麼？
4. 如何讓 print() 不換行？
5. 如何獲取使用者輸入的整數？

**參考答案**：
1. 空格（`' '`）
2. str（字串）
3. `3.14`
4. `print(..., end="")`
5. `num = int(input())`

---

### 🔗 延伸閱讀（Further Reading）

#### Python 官方文件
- [Input and Output](https://docs.python.org/3/tutorial/inputoutput.html)
- [Format String Syntax](https://docs.python.org/3/library/string.html#format-string-syntax)

#### 推薦資源
- [Real Python: Python String Formatting](https://realpython.com/python-f-strings/)
- [PyFormat](https://pyformat.info/)

#### 下一步
- **Unit 4: 條件判斷** - 根據輸入做決策

---

## 💪 即時練習（Quick Practice）

請在下方 cell 完成以下任務：

1. 提示使用者輸入姓名、年齡、身高(cm)
2. 計算 5 年後的年齡
3. 將身高轉換為公尺
4. 用格式化輸出顯示結果（小數保留 2 位）

In [13]:
# 在此撰寫你的程式碼
# TODO: 請填入適當的運算式並取消註解

name = input("姓名：")
age = int(input("年齡："))
height_cm = float(input("身高(cm)："))

# 計算
# future_age = age + 5  # 請完成：5年後的年齡
# height_m = height_cm / 100  # 請完成：將公分轉換為公尺

# 輸出（使用 f-string 格式化）
# print(f"\n{name}，你現在 {age} 歲，5年後 {future_age} 歲")
# print(f"身高：{height_m:.2f} 公尺")